TR1: add quadrilateral interpolation

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

Resolves #354.
![output](https://github.com/user-attachments/assets/e64139af-45c8-4480-b9f6-5f021b9403e0)

#### Testing

It requires testing TR2 SWR/HWR × ZBuffer On/Off as well, since this heavily touches the common code.

rr- committed:
output: move rhw multiplication to the shader

rr- committed:
tr1/output: add quadrilateral interpolation

Resolves #354

walkawayy commented on src/tr2/game/game_string.def:
```diff
@@ -39,3 +39,5 @@ GS_DEFINE(OSD_LIGHTNING_CONTRAST_FMT, "Lighting Contrast: %s")
 GS_DEFINE(OSD_HARDWARE_RENDERING, "Hardware rendering")
 GS_DEFINE(OSD_SOFTWARE_RENDERING, "Software rendering")
+GS_DEFINE(OSD_PERSPECTIVE_FILTER_ON, "Perspective correction: on")
+GS_DEFINE(OSD_PERSPECTIVE_FILTER_OFF, "Perspective correction: off")
```
Is there a reason it's labelled something different in TR2? I also didn't notice any different with it on vs off.

rr- replied:
Linter complains that it makes no sense to define this string in trx if it's only used in tr2 now.

walkawayy reviewed (changes requested):
Looks really good. I posted some screenshots in Discord comparing. The only downgrade I noticed is the Sanctuary of the Scion sphinx face. (upload size too big for Github)

The only bug I noticed is the graphics menu text overlaps.

![20250309_175959_City_of_Vilcabamba](https://github.com/user-attachments/assets/0aa932dd-72a6-49fe-8a9b-5fd8ac41bbbf)

lahm86 commented:
LGTM. Just a couple of points that have come to mind (and I'm happy to raise these separately if best, and/or assist with solutions).

- I have a habit of using Alt-F4 to exit the game, and currently this toggles the setting

- In TombEditor, you can create textures like this, so where you want a clean cut with a non-standard shape, but the result in-game isn't ideal with the setting on. Perhaps we could introduce a way for builders to force the setting off. Our gameflow enforced approach doesn't apply to in-game togglable settings (so this is perhaps best separate, as that's a flaw in the enforced logic ultimately).
![image](https://github.com/user-attachments/assets/3e167245-be72-4a3c-aff8-b5a34e2caea0)

![image](https://github.com/user-attachments/assets/d8254832-ed0b-49fa-a2f2-9e5576cb72ec)

Richard-L commented:
Thank you so so so much for tackling this 🙏
In some areas it means a dramatic improvement (gym blocks, rock textures, vilcabamba chairs, etc).

Some things I noticed:
1. There's a strange black outline around text in the passport, and 2D sprites:
  
![image](https://github.com/user-attachments/assets/1ee3063d-6f83-49a6-b642-49eed429375f)
![image](https://github.com/user-attachments/assets/b57d09fe-57a5-4f98-a8f1-9bfacd9d03ee)

2. The backdrop garden in Lara's home is missing
![20250309_232522_Laras_Home](https://github.com/user-attachments/assets/cc612f1f-f733-4463-96e9-f178e3aaf914)

3. The pool room in the gym is too (?)
![20250309_232648_Laras_Home](https://github.com/user-attachments/assets/134f45e5-a867-4595-8299-a46d154966ed)

4. In some levels, certain things are invisible:
- parts of Lara
- blood sprites
- darts
![20250309_231257_Tomb_of_Tihocan](https://github.com/user-attachments/assets/3bbf72bc-52bd-41f3-b478-1525b72c1ebc)
![20250309_231333_Tomb_of_Qualopec](https://github.com/user-attachments/assets/d7bacb76-81bf-4d1e-a4f7-900165b45599)

walkawayy commented:
@Richard-L did you update your `shaders` folder with the files from this PR? I didn't at first, and it caused weird bugs like this. The gardens and the pool room in the gym look fine for me.

rr- committed:
input: prevent alt-f4 from being passed to game

rr- commented:
- Fixed the Alt-F4 problem
- Fixed the dialog sizing problem
- Awaiting a test level so that I can try fix the non-rectilinear UV meshes problem
- Not fixing the Sphinx face atm, I'd need assistance with this

aredfan commented:
It's very cool to see work on quadrilateral interpolation is underway. 👍 

I'm getting the same issue as Richard but it only happens when bilinear filter is enabled. I also replaced the shader files as suggested by walkawayy, with the ones included with the PR build.

rr- commented:
Thanks – fixed these as well. Fortunately it was a small typo in the first refactor-only commit, rather than a problem with the algorithm.

aredfan commented:
I came across 2 potential texture issues. Left is with trapezoid disabled, right is with said option enabled.

- [x] Lara's left eye when she is being eaten by the T-rex.

![1](https://github.com/user-attachments/assets/63049abd-81fa-40cb-ba9e-7a64a1bb0c85)

- [x] I think Natla's eyes are a bit wide apart - `/cut 4`, then enter photo mode to get close to her face then press F4.

![2](https://github.com/user-attachments/assets/0e0de5e9-eca8-4cb6-a85f-0ad53867f38d)

Natla's model is present in the following levels:

- CUT3.PHD
- CUT4.PHD
- LEVEL10C.PHD

The other humans in the game LGTM.

rr- committed:
tr1/output: improve trapezoid transform

rr- commented:
I think it'll require injecting different mappings of UV meshes. Same with the Sphinx model.

rr- committed:
tr1/output: fix resetting trapezoid ratios

rr- commented:
@aredfan fixed Natla, Lara's eye and Sphinx. It was an issue with the code after all which didn't reset the extra parameters for the triangle buffer.

aredfan commented:
> @aredfanfixed Natla, Lara's eye and Sphinx. It was an issue with the code after all which didn't reset the extra parameters for the triangle buffer.

Glad to hear it. I can confirm all of these issues are fixed. :+1:
